From 4a8f4fd1c48d5c1493fc445506a9bba85fc1016f Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Mon, 30 Jan 2023 18:26:10 +0100
Subject: [PATCH 0247/1695] drivers: stm32_rifsc: add lock capabilities for the
 RIFSC

Adds support for the st,glocked property which is used for the global
lock of the RIF configuration of RIMUs/RISUPs. It is possible to
globally lock either RIMUs, RISUPs or both.

RISUPs support peripheral granularity for configuration locking. Adds
support for this feature in RIFPROT binding at offset 10.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 68ff3ce96a263efd07746157fd302ab66bf9b600 drivers: stm32_rifsc: add lock capabilities for the RIFSC]
Change-Id: I463b757e27993e31355d2ce7bb1301a0cc2a809d
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
---
 core/drivers/firewall/stm32_rifsc.c | 64 +++++++++++++++++++++++++++--
 core/include/drivers/stm32_rifsc.h  |  3 +-
 2 files changed, 62 insertions(+), 5 deletions(-)

diff --git a/core/drivers/firewall/stm32_rifsc.c b/core/drivers/firewall/stm32_rifsc.c
index 7fbe736ac..bcb255f8a 100644
--- a/core/drivers/firewall/stm32_rifsc.c
+++ b/core/drivers/firewall/stm32_rifsc.c
@@ -21,6 +21,7 @@
 /* RIFSC offset register */
 #define _RIFSC_RISC_SECCFGR0		U(0x10)
 #define _RIFSC_RISC_PRIVCFGR0		U(0x30)
+#define _RIFSC_RISC_RCFGLOCKR0		U(0x50)
 #define _RIFSC_RISC_PER0_CIDCFGR	U(0x100)
 #define _RIFSC_RIMC_CR			U(0xC00)
 #define _RIFSC_RIMC_ATTR0		U(0xC10)
@@ -52,9 +53,15 @@
 #define _RIFSC_HWCFGR1_CFG6_MASK	GENMASK_32(23, 20)
 #define _RIFSC_HWCFGR1_CFG6_SHIFT	U(20)
 
+/*
+ * RISC_CR register fields
+ */
+#define _RIFSC_RISC_CR_GLOCK		BIT(0)
+
 /*
  * RIMC_CR register fields
  */
+#define _RIFSC_RIMC_CR_GLOCK		BIT(0)
 #define _RIFSC_RIMC_CR_TDCID_MASK	GENMASK_32(6, 4)
 
 /* RIFSC_VERR register fields */
@@ -158,6 +165,46 @@ static void stm32_rifsc_get_driverdata(struct rifsc_platdata *pdata)
 	     rifsc_drvdata.nb_risal);
 }
 
+static TEE_Result stm32_rifsc_glock_config(const void *fdt, int node,
+					   struct rifsc_platdata *pdata)
+{
+	const fdt32_t *cuint = NULL;
+	int len = 0;
+	uint32_t glock_conf = 0;
+
+	cuint = fdt_getprop(fdt, node, "st,glocked", &len);
+	if (!cuint) {
+		DMSG("No global lock on RIF configuration");
+		return TEE_SUCCESS;
+	}
+
+	assert(len == sizeof(uint32_t));
+
+	glock_conf = fdt32_to_cpu(*cuint);
+
+	if (glock_conf & RIFSC_RIMU_GLOCK) {
+		DMSG("Setting global lock on RIMU configuration");
+
+		io_setbits32(pdata->base + _RIFSC_RIMC_CR,
+			     _RIFSC_RIMC_CR_GLOCK);
+
+		if (!(io_read32(pdata->base + _RIFSC_RIMC_CR) &
+		      _RIFSC_RIMC_CR_GLOCK))
+			return TEE_ERROR_ACCESS_DENIED;
+	}
+
+	if (glock_conf & RIFSC_RISUP_GLOCK) {
+		DMSG("Setting global lock on RISUP configuration");
+
+		io_setbits32(pdata->base, _RIFSC_RISC_CR_GLOCK);
+
+		if (!(io_read32(pdata->base) & _RIFSC_RISC_CR_GLOCK))
+			return TEE_ERROR_ACCESS_DENIED;
+	}
+
+	return TEE_SUCCESS;
+}
+
 static TEE_Result stm32_rifsc_dt_conf_risup(const void *fdt, int node,
 					    int *nrisup,
 					    struct risup_cfg **risups)
@@ -185,6 +232,7 @@ static TEE_Result stm32_rifsc_dt_conf_risup(const void *fdt, int node,
 		risup->id = _RIF_FLD_GET(RIFSC_RISC_PER_ID, value);
 		risup->sec = (bool)_RIF_FLD_GET(RIFSC_RISC_SEC, value);
 		risup->priv = (bool)_RIF_FLD_GET(RIFSC_RISC_PRIV, value);
+		risup->lock = (bool)_RIF_FLD_GET(RIFSC_RISC_LOCK, value);
 		risup->cid_attr = _RIF_FLD_GET(RIFSC_RISC_PERx_CID, value);
 	}
 
@@ -354,6 +402,7 @@ static TEE_Result stm32_risup_cfg(struct rifsc_platdata *pdata,
 				  struct risup_cfg *risup)
 {
 	struct rifsc_driver_data *drv_data = pdata->drv_data;
+	uintptr_t cidcfgr_offset = _OFST_PERX_CIDCFGR * risup->id;
 	uintptr_t offset = sizeof(uint32_t) * (risup->id / _PERIPH_IDS_PER_REG);
 	uint32_t shift = risup->id % _PERIPH_IDS_PER_REG;
 
@@ -368,10 +417,15 @@ static TEE_Result stm32_risup_cfg(struct rifsc_platdata *pdata,
 		io_clrsetbits32(pdata->base + _RIFSC_RISC_PRIVCFGR0 + offset,
 				BIT(shift), risup->priv << shift);
 
-	if (drv_data->rif_en) {
-		offset = _OFST_PERX_CIDCFGR * risup->id;
-		io_write32(pdata->base + _RIFSC_RISC_PER0_CIDCFGR + offset,
-			   risup->cid_attr);
+	if (drv_data->rif_en)
+		io_write32(pdata->base + _RIFSC_RISC_PER0_CIDCFGR +
+			   cidcfgr_offset, risup->cid_attr);
+
+	/* Lock configuration for this RISUP */
+	if (risup->lock) {
+		DMSG("Locking RIF conf for peripheral n°%"PRIu32, risup->id);
+		io_setbits32(pdata->base + _RIFSC_RISC_RCFGLOCKR0 + offset,
+			     BIT(shift));
 	}
 
 	return TEE_SUCCESS;
@@ -511,6 +565,8 @@ static TEE_Result stm32_rifsc_probe(const void *fdt, int node,
 	if (res)
 		return res;
 
+	stm32_rifsc_glock_config(fdt, node, &rifsc_pdata);
+
 	fdev = stm32_firewall_dev_alloc();
 	if (!fdev)
 		return TEE_ERROR_OUT_OF_MEMORY;
diff --git a/core/include/drivers/stm32_rifsc.h b/core/include/drivers/stm32_rifsc.h
index 03d448759..045283a54 100644
--- a/core/include/drivers/stm32_rifsc.h
+++ b/core/include/drivers/stm32_rifsc.h
@@ -9,10 +9,11 @@
 #include <util.h>
 
 struct risup_cfg {
+	uint32_t cid_attr;
 	uint32_t id;
 	bool sec;
 	bool priv;
-	uint32_t cid_attr;
+	bool lock;
 };
 
 struct rimu_cfg {
-- 
2.47.2

